### ISSUE TITLE
Unexpected Behavior in InfoNCE Loss Implementation for Qwen Embedding Training

### ISSUE THREAD

**Issue by Alireza3242** [modelscope/ms-swift#6969]
**Describe the bug**

I'm training a Qwen embedding model using the InfoNCE loss function and have encountered what seems to be a bug or unintended behavior in its implementation.

Let’s define the batch components: queries as q[0], q[1], ..., q[B-1], positive documents as p[0], p[1], ..., p[B-1], and negative documents as n[0][0:k], ..., n[B-1][0:k]. So, each query has k associated negative documents.

Issue 1: INFONCE_USE_BATCH Flag Behaves Incorrectly
When I set INFONCE_USE_BATCH = True, the loss function incorrectly computes the similarity between q[i] and n[j][0:k]—the negative documents from other samples in the batch.
The expected behavior is that when this flag is enabled, q[i] should only be compared to the positive documents from other batch samples. It should not involve negatives from other queries. However, the current implementation seems to incorporate both positive and negative documents from all batch samples, which appears to be a bug.

Issue 2: INFONCE_MASK_FAKE_NEGATIVE Incorrectly Masks Hard Negatives
When I enable INFONCE_MASK_FAKE_NEGATIVE, the function applies its "fake negative" masking check even to my hard negatives.
This is problematic because:

- I have explicitly curated these hard negatives. While they are challenging for the model, I am certain they are true negatives.

- In contrast, I am not certain if documents from other random samples in the batch are truly negative. The masking logic is intended for this latter, uncertain case.
    The ideal (and expected) behavior is for the masking check to be applied only to documents from other batch samples, not to the pre-defined hard negatives. Including hard negatives in this mask weakens the training signal.

In addition to the functional bugs previously mentioned, there is a significant issue regarding code maintainability and understandability.
The function implementing the InfoNCE loss is very long and complex. This makes it difficult to read, debug, and modify.
To improve code clarity and reduce its cognitive complexity (as measured by tools like SonarQube), this monolithic function should be refactored into smaller, well-named, and focused helper functions. These helper functions would then be called within a main, high-level compute_infonce_loss function.

**Comment by tastelikefeet** [comment]
Sorry for late replying.
For the 1st issue: I think this behaviour is correct.  Here is the original paper:
https://arxiv.org/pdf/2506.05176

<img width="1824" height="1266" alt="Image" src="https://github.com/user-attachments/assets/de41be5f-31c1-4bb2-8fba-0355e8767c48" />
So I think the negative samples of other queries are included.

For the 2nd issue:
From the paper's opinion, the implementation possibly did not exclude the negative parts of one sample itself. However, you can exclude its own negative parts by:
```
if infonce_mask_fake_negative:
    ...
    # The original qd_mask line
    qd_mask = qd_block > thresholds
    batch_size = sentences.size(0)
    block_size = sentences.size(1) - 1  # (neg+1)
    own_block_mask = torch.zeros_like(qd_mask, dtype=torch.bool)
    for i in range(batch_size):
        start_idx = i * block_size
        end_idx = (i + 1) * block_size
        own_block_mask[i, start_idx:end_idx] = True
    # The original qd_mask with extra own_block_mask
    qd_mask = qd_mask & (~own_block_mask)
    ...
```

For the 3rd issue:
The infonce loss is quite long because:
1. The loss is compatible with scenarios of length-equal samples and length-non-equal samples, so there is a `can_batched` parameter
2. The loss function varies from gte models/gme models to qwen3 embedding models, and some requirements from other uses, so there are some compatible code

Sorry for bring you inconvenience, if you have any suggestions to split this function to independent loss functions, please feel free to give us comments or PRs
